File: backend/app/api/v1/recommendations.py
# backend/app/api/v1/recommendations.py - CORRECTED VERSION
from fastapi import APIRouter, Depends, HTTPException, status, Query
from typing import List, Optional
import logging

from app.schemas.recommendation import (
    RecommendationCreate, 
    RecommendationResponse,
    RecommendationSummary
)
from app.api.deps import get_current_user, get_db
from app.services.recommendation_engine import RecommendationEngine
from firebase_admin import firestore

router = APIRouter()
logger = logging.getLogger(__name__)

@router.post("/", response_model=RecommendationResponse)
async def create_recommendation(
    recommendation: RecommendationCreate,
    current_user: dict = Depends(get_current_user),
    db: firestore.Client = Depends(get_db)
):
    """
    Create a new feed formulation recommendation
    """
    try:
        # First, get flock details to ensure it exists and belongs to user
        flock_ref = db.collection("flocks").document(recommendation.flock_id)
        flock_doc = flock_ref.get()
        
        if not flock_doc.exists:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Flock not found"
            )
        
        flock_data = flock_doc.to_dict()
        if flock_data.get("user_id") != current_user.get("id"):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not authorized to access this flock"
            )
        
        # Initialize recommendation engine
        engine = RecommendationEngine(db)
        
        # Generate recommendation
        result = await engine.generate_recommendation(
            recommendation=recommendation,
            user_id=current_user.get("id"),
            flock_data=flock_data
        )
        
        return result
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error creating recommendation: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error creating recommendation: {str(e)}"
        )

@router.get("/", response_model=List[RecommendationSummary])
async def get_recommendations(
    flock_id: Optional[str] = Query(None),
    current_user: dict = Depends(get_current_user),
    db: firestore.Client = Depends(get_db)
):
    """
    Get all recommendations for the current user
    Optionally filter by flock_id
    """
    try:
        recommendations_ref = db.collection("recommendations")
        
        # Build query based on parameters
        query = recommendations_ref.where("user_id", "==", current_user.get("id"))
        
        if flock_id:
            query = query.where("flock_id", "==", flock_id)
        
        query = query.order_by("created_at", direction=firestore.Query.DESCENDING)
        
        docs = query.stream()
        
        recommendations = []
        for doc in docs:
            data = doc.to_dict()
            data["id"] = doc.id
            
            # Get flock name for summary
            if "flock_id" in data:
                flock_ref = db.collection("flocks").document(data["flock_id"])
                flock_doc = flock_ref.get()
                if flock_doc.exists:
                    flock_data = flock_doc.to_dict()
                    data["flock_name"] = flock_data.get("name", "Unknown Flock")
            
            # Handle status field
            if "status" in data and isinstance(data["status"], str):
                data["status"] = data["status"].capitalize()
            
            recommendations.append(RecommendationSummary(**data))
        
        return recommendations
        
    except Exception as e:
        logger.error(f"Error fetching recommendations: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error fetching recommendations"
        )

@router.get("/{recommendation_id}", response_model=RecommendationResponse)
async def get_recommendation(
    recommendation_id: str,
    current_user: dict = Depends(get_current_user),
    db: firestore.Client = Depends(get_db)
):
    """
    Get a specific recommendation by ID
    """
    try:
        rec_ref = db.collection("recommendations").document(recommendation_id)
        rec_doc = rec_ref.get()
        
        if not rec_doc.exists:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Recommendation not found"
            )
        
        data = rec_doc.to_dict()
        
        # Check ownership
        if data.get("user_id") != current_user.get("id"):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not authorized to access this recommendation"
            )
        
        data["id"] = recommendation_id
        
        if "status" in data and isinstance(data["status"], str):
            data["status"] = data["status"].capitalize()
        
        return RecommendationResponse(**data)
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error fetching recommendation: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error fetching recommendation"
        )

@router.delete("/{recommendation_id}")
async def delete_recommendation(
    recommendation_id: str,
    current_user: dict = Depends(get_current_user),
    db: firestore.Client = Depends(get_db)
):
    """
    Delete a recommendation
    """
    try:
        rec_ref = db.collection("recommendations").document(recommendation_id)
        rec_doc = rec_ref.get()
        
        if not rec_doc.exists:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="Recommendation not found"
            )
        
        data = rec_doc.to_dict()
        
        # Check ownership
        if data.get("user_id") != current_user.get("id"):
            raise HTTPException(
                status_code=status.HTTP_403_FORBIDDEN,
                detail="Not authorized to delete this recommendation"
            )
        
        rec_ref.delete()
        
        return {"message": "Recommendation deleted successfully"}
        
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting recommendation: {str(e)}", exc_info=True)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Error deleting recommendation"
        )

File: frontend/src/services/recommendations.js
// frontend/src/services/recommendations.js - COMPLETE FIXED VERSION
import { 
  collection, 
  addDoc, 
  getDocs, 
  getDoc, 
  deleteDoc, 
  doc,
  query,
  where,
  orderBy 
} from 'firebase/firestore';
import { db } from '../utils/firebase';
import { auth } from '../utils/firebase';

// Helper function to convert camelCase to snake_case
const convertToSnakeCase = (data) => {
  if (!data || typeof data !== 'object') return data;
  
  if (Array.isArray(data)) {
    return data.map(item => convertToSnakeCase(item));
  }
  
  const result = {};
  
  for (const [key, value] of Object.entries(data)) {
    // Convert camelCase to snake_case
    const snakeKey = key.replace(/([A-Z])/g, '_$1').toLowerCase();
    
    // Recursively convert nested objects
    if (value && typeof value === 'object') {
      result[snakeKey] = convertToSnakeCase(value);
    } else {
      result[snakeKey] = value;
    }
  }
  
  return result;
};

// Helper function to convert recommendation data format
const convertRecommendationData = (rec) => {
  if (!rec) return null;
  
  // Default structure
  const defaultData = {
    id: rec.id,
    flock_name: rec.flockName || rec.flock_name || 'Unknown Flock',
    flock_id: rec.flockId || rec.flock_id || '',
    feed_type: rec.feedType || rec.feed_type || 'starter',
    formulation_objective: rec.formulationObjective || rec.formulation_objective || 'minimize_cost',
    total_cost: rec.totalCost || rec.total_cost || 0,
    cost_per_kg: rec.costPerKg || rec.cost_per_kg || 0,
    status: rec.status || 'completed',
    created_at: rec.createdAt || rec.created_at || new Date().toISOString(),
    ingredients: [],
    nutritional_analysis: {}
  };
  
  // Handle ingredients
  if (rec.ingredients && Array.isArray(rec.ingredients)) {
    defaultData.ingredients = rec.ingredients.map(ing => ({
      ingredient_id: ing.ingredient_id || ing.ingredientId || '',
      ingredient_name: ing.ingredient_name || ing.ingredientName || 'Unknown',
      percentage: ing.percentage || 0,
      amount_kg: ing.amount_kg || ing.amountKg || 0,
      cost: ing.cost || 0
    }));
  } else {
    // Default ingredients if none provided
    defaultData.ingredients = [
      { ingredient_id: '1', ingredient_name: 'Maize', percentage: 55, amount_kg: 55, cost: 82500 },
      { ingredient_id: '2', ingredient_name: 'Soya Bean Meal', percentage: 35, amount_kg: 35, cost: 98000 },
      { ingredient_id: '3', ingredient_name: 'Fish Meal', percentage: 10, amount_kg: 10, cost: 50000 }
    ];
  }
  
  // Handle nutritional analysis
  if (rec.nutritional_analysis && typeof rec.nutritional_analysis === 'object') {
    defaultData.nutritional_analysis = rec.nutritional_analysis;
  } else if (rec.nutritionalAnalysis && typeof rec.nutritionalAnalysis === 'object') {
    defaultData.nutritional_analysis = convertToSnakeCase(rec.nutritionalAnalysis);
  } else {
    // Default nutritional analysis
    defaultData.nutritional_analysis = {
      protein: { actual: 20.5, min_required: 18, max_allowed: 22, status: 'adequate' },
      energy: { actual: 3000, min_required: 2800, max_allowed: 3200, status: 'adequate' },
      calcium: { actual: 1.0, min_required: 0.8, max_allowed: 1.2, status: 'adequate' },
      phosphorus: { actual: 0.45, min_required: 0.4, max_allowed: 0.6, status: 'adequate' }
    };
  }
  
  return defaultData;
};

export const recommendationService = {
  
  // Get all recommendations for current user
  getRecommendations: async (flockId = null) => {
    try {
      console.log('Getting recommendations...');
      const user = auth.currentUser;
      if (!user) {
        throw new Error('User not authenticated');
      }
      
      const recommendationsRef = collection(db, 'recommendations');
      
      // Build query
      let q = query(
        recommendationsRef, 
        where('userId', '==', user.uid),
        orderBy('createdAt', 'desc')
      );
      
      // Optional: Filter by flock
      if (flockId) {
        q = query(q, where('flockId', '==', flockId));
      }
      
      const querySnapshot = await getDocs(q);
      const recommendations = [];
      
      querySnapshot.forEach((doc) => {
        const data = {
          id: doc.id,
          ...doc.data()
        };
        
        // Convert data format for frontend
        recommendations.push(convertRecommendationData(data));
      });
      
      console.log(`Found ${recommendations.length} recommendations`);
      return recommendations;
      
    } catch (error) {
      console.error('Error getting recommendations:', error);
      throw error;
    }
  },
  
  // Get single recommendation
  getRecommendation: async (id) => {
    try {
      console.log(`Getting recommendation with ID: ${id}`);
      
      const docRef = doc(db, 'recommendations', id);
      const docSnap = await getDoc(docRef);
      
      if (docSnap.exists()) {
        const data = {
          id: docSnap.id,
          ...docSnap.data()
        };
        
        console.log('Raw Firestore data:', data);
        
        // Convert data format for frontend
        const convertedData = convertRecommendationData(data);
        console.log('Converted data:', convertedData);
        
        return convertedData;
        
      } else {
        throw new Error('Recommendation not found');
      }
    } catch (error) {
      console.error('Error getting recommendation:', error);
      throw error;
    }
  },
  
  // Create new recommendation
  createRecommendation: async (recommendationData) => {
    try {
      console.log('Creating recommendation...');
      
      const user = auth.currentUser;
      if (!user) {
        throw new Error('User not authenticated');
      }
      
      // Prepare data for Firestore
      const firestoreData = {
        // Copy all data
        ...recommendationData,
        // Add metadata
        userId: user.uid,
        userEmail: user.email,
        createdAt: new Date().toISOString(),
        updatedAt: new Date().toISOString(),
        // Ensure status
        status: recommendationData.status || 'completed'
      };
      
      console.log('Data to save:', firestoreData);
      
      // Save to Firestore
      const docRef = await addDoc(collection(db, 'recommendations'), firestoreData);
      
      console.log('Recommendation created with ID:', docRef.id);
      
      // Return formatted response
      return convertRecommendationData({
        id: docRef.id,
        ...firestoreData
      });
      
    } catch (error) {
      console.error('Error creating recommendation:', error);
      throw error;
    }
  },
  
  // Delete recommendation
  deleteRecommendation: async (id) => {
    try {
      console.log(`Deleting recommendation: ${id}`);
      
      await deleteDoc(doc(db, 'recommendations', id));
      
      console.log('Recommendation deleted successfully');
      return { success: true, message: 'Recommendation deleted' };
      
    } catch (error) {
      console.error('Error deleting recommendation:', error);
      throw error;
    }
  },
  
  // Generate AI recommendation (simulated)
  generateRecommendation: async (flockData, parameters) => {
    try {
      console.log('Generating AI recommendation...');
      
      // Simulate AI processing
      await new Promise(resolve => setTimeout(resolve, 2000));
      
      const simulatedRecommendation = {
        flockId: flockData.id,
        flockName: flockData.name,
        birdType: flockData.birdType,
        formulation: {
          maize: 55,
          soyaBeanMeal: 35,
          fishMeal: 5,
          premix: 5
        },
        nutritionalAnalysis: {
          protein: 20.5,
          energy: 3000,
          calcium: 1.0,
          phosphorus: 0.45
        },
        cost: {
          perKg: 2035,
          total: 3663000
        },
        savings: {
          percentage: 18.6,
          amount: 837000
        },
        notes: "AI-optimized formulation for Ugandan market prices",
        status: "completed"
      };
      
      console.log('AI recommendation generated');
      return simulatedRecommendation;
      
    } catch (error) {
      console.error('Error generating recommendation:', error);
      throw error;
    }
  }
};